### DISCUSSION TITLE
ClusterData Parent Class

### DISCUSSION THREAD

**Discussion by suxuann** [pyg-team/pytorch_geometric#2243] [Q&A]
The core functionality of ClusterData is to partition a large graph into multiple subgraphs.

My question is: Why does it subclass from the generic torch.utils.data.Dataset, and not torch_geometric.data.dataset.Dataset?

The rationale for subclassing from the latter class is that we can view a dataset consisting of subgraphs as a new dataset altogether, and use some of its functionalities (although admittedly, there are some differences).

I mostly followed the example in **https://github.com/rusty1s/pytorch_geometric/blob/master/examples/cluster_gcn_reddit.py**

Additionally, I bumped into a problem using ClusterLoader while using a for-loop over it. It seems like it only has one element even though I have specified num_parts in ClusterData.

**Comment by suxuann** [comment]
Moreover, the original paper contains methods to group clusters, and use these for training. This seems to be lacking in the current implementation. Would it be useful?

**Comment by rusty1s** [comment]
While one can view `ClusterData` as a dataset/list of subgraphs, it is not recommended in doing so because you will lose out on the inter-connectivity between clusters when iterating over it in mini-batches. The `ClusterData` is more meant as a simple interface for inspecting different subgraphs, while it is recommended to use `ClusterLoader` to iterate over it.

I think one could also implement it as a subclass of `torch_geometric.data.Dataset`, but we do not need the `download` and `process` functionality that the PyG dataset interface provides.
